exam (#36)
romankuchin/ica0002-2021: exam/meta.md

Exam related information
------------------------

Exam task: [README.md](./README.md)

Access
------

Exam times are listed on the [main course page](../README.md).

You can choose any 2 times to take the exam in addition to week 16 exam attempt,
3 attempts in total.

To get access to the exam you need to:

1. Complete all the 14 lab tasks
2. Get your solution verified by Judge Dredd:
    http://193.40.156.86/students.html > your-name > Results should be all green

Once done, you will soon get a third virtual machine to prepare and test your
solution.

Rules
-----

Exam starts strictly at the announced time and lasts for exactly
**3 hours and 15 minutes** (195 minutes) -- not a minute longer.

Submissions are checked on first-come-first-served basis.

This is open-everything exam: you can use whatever materials you want while
writing the code, testing it or presenting.

Pre-testing
-----------

> Before taking the exam you should verify your solution by yourselves.

Reviewer: ```suggestion
Before taking the exam you should verify your solution yourself.
```

> Run this script `pre-exam.sh` from your repo root on *empty VMs*, it will create a file `pre-exam.txt`, commit it to main branch. Just a reminder, that commiting your passwords is bad idea. If playbook execution failed - fix your code, recreate VMs and try again.

Reviewer: ```suggestion
Run this script `pre-exam.sh` from your repo root on *empty VMs*. It will create a file `pre-exam.txt` -- commit it to `main` branch. Just a reminder: commiting your passwords is a bad idea. If playbook execution failed -- fix your code, recreate VMs and try again.
```

Expected result:

- Many changes on 1st run
- Zero changes on 2nd run
- Zero changes on 3rd run

> If you have something different - fix your code. No point to take the exam until you have

Reviewer: ```suggestion
If you have something different -- fix your code. No point to take the exam until you have achieved the
```

> expected result, we will check your playbooks during the exam anyway.

Reviewer: ```suggestion
expected result. Note that we will still check your code during the exam.
```

Testing
-------

This is what we (teachers) may do to test your infrastructure.

**1. Kill any of the services mentined above**

You should be able to restore the infrastructure in exactly one Ansible run.

Note: to _restore_ the services with Ansible here and in other cases you may --
but don't have to -- use Ansible tags.

**2. Kill one of the highly available service instances**

Agama in Docker container, Bind (master or slave), Docker daemon, HAProxy,

> Keepalived, MySQL.

Reviewer: ```suggestion
Keepalived, MySQL replica.
```
